src/shogun/multiclass/KNNSolver.h

```diff
@@ -0,0 +1,95 @@
+/* This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Copyright (c) 2012-2013 Sergey Lisitsyn
+ */
+
+#ifndef KNNSOLVER_H__
+#define KNNSOLVER_H__
+
+#include <shogun/lib/config.h>
+
+#include <shogun/lib/common.h>
+#include <shogun/io/SGIO.h>
+#include <shogun/distance/Distance.h>
+#include <shogun/machine/DistanceMachine.h>
+
+namespace shogun
+{
+/**
+ * The KNNSolver class is the parent class of rest of other solver class. The KNNSolver class will 
+ * include most of Variables which been used all the other Solver and methods been used to choose 
+ * the index of the most frequent class.
+ */
+
+class CDistanceMachine;
+class CKNNSolver : public CDistanceMachine
+{
+	public:
+		/** default constructor */
+		CKNNSolver(): CDistanceMachine() { init(); }
+
+		virtual ~CKNNSolver() { /* nothing to do */ }
+
+		CKNNSolver(const int32_t k, const float64_t q, const int32_t num_classes, const int32_t min_label, const SGVector<int32_t> train_labels);
+
+		/** compute the histogram of class outputs of the k nearest 
+		 * neighbors to a test vector and return the index of the most frequent class
+		 *
+		 * @param classes vector used to store the histogram
+		 * @param train_lab class indices of the training data. If the cover
+		 * tree is not used, the elements are ordered by increasing distance
+		 * and there are elements for each of the training vectors. If the cover
+		 * tree is used, it contains just m_k elements not necessary ordered.+		 *
+		 * @return index of the most frequent class, class detected by KNN
+		 */
+		const int32_t choose_class(float64_t* classes, const int32_t* train_lab);
```

Comment:
why am i right that this function does not change the object itself (parameters of it)?
because then you can define the whole function to be const...